[BUGFIX] Fix `willDestroy` on class helpers

pzuraq opened the pull request:
Class based helpers were having `willDestroy` be called twice, once by
the Glimmer VM's built-in destroyables implementation, and once
scheduled by `destroy` (which is the correct destruction). This is
because we were registering the class itself directly as the
destroyable, and Glimmer VM detected the `willDestroy` function on it
and called it eagerly.

This fix creates an intermediate destructor object with a `destroy()`
hook, but long term we need to finish the refactors to destroyables in
the VM and make it so these types of collisions can't happen again.

Added a basic lifecycle test for helpers to prevent this specific issue
from reoccuring.

Fixes #18830

rwjblue commented:
Can you update this to `[BUGFIX release]`?

rwjblue commented on yarn.lock:
```diff
@@ -2,7 +2,7 @@
 
 
-"@babel/code-frame@^7.0.0", "@babel/code-frame@^7.5.5", "@babel/code-frame@^7.8.3":
+"@babel/code-frame@^7.0.0", "@babel/code-frame@^7.8.3":
```
Can you reset this files changes? Seems unrelated to the other changes here...

pzuraq replied:
FWIW I think this just happened when I ran `yarn` locally`, unsure. But yes, will reset

Commit:
[BUGFIX release] Fix `willDestroy` on class helpers #18837

Class based helpers were having `willDestroy` be called twice, once by
the Glimmer VM's built-in destroyables implementation, and once
scheduled by `destroy` (which is the correct destruction). This is
because we were registering the class itself directly as the
destroyable, and Glimmer VM detected the `willDestroy` function on it
and called it eagerly.

This fix creates an intermediate destructor object with a `destroy()`
hook, but long term we need to finish the refactors to destroyables in
the VM and make it so these types of collisions can't happen again.

Added a basic lifecycle test for helpers to prevent this specific issue
from reoccuring.
